### ISSUE TITLE
R package failed R CMD check under Windows for arch 'i386'

### ISSUE THREAD

**Issue by hetong007** [dmlc/xgboost#2319]
## Environment info
Operating System: Windows 10

`xgboost` version used: 29289d2

The R `sessionInfo()`:
```
> sessionInfo()
R version 3.4.0 (2017-04-21)
Platform: i386-w64-mingw32/i386 (32-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=Chinese (Simplified)_China.936 
[2] LC_CTYPE=Chinese (Simplified)_China.936   
[3] LC_MONETARY=Chinese (Simplified)_China.936
[4] LC_NUMERIC=C                              
[5] LC_TIME=Chinese (Simplified)_China.936    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.0 tools_3.4.0
```

## Steps to reproduce

1. (under Linux/OSX) run `make Rbuild` to generate the tarball
2. Under Windows with R 3.4.0 and both x64/i386 installed, run 
`R CMD check --multiarch xgboost_0.6.4.4.tar.gz`
3. When checking examples and tests for arch `i386`, there will be errors with popup windows (look like corrupted memory). No cause of error in the print-out messages from `R CMD check`.

## What have you tried?

1. Remove all files under `tests/testthat` folder except `test_basic.R`, then simplify its content (no multithreading, no testthat wrapper, no sparse matrix):
```
require(xgboost)

train = matrix(10,5,2)
label = c(0,1,0,0,1)

bst <- xgboost(data = train, label = label, max_depth = 2,
               eta = 1, nthread = 1, nrounds = 2, objective = "binary:logistic")
```
The error remains. So the `xgboost()` call triggered the error. 
2. However, if replace the `xgboost()` call with `xgb.DMatrix`, then there's no error triggered. 
3. Not sure how to reproduce out of the `R CMD check` yet. Tried `gctorture(TRUE)` but with no error.

Could you guys please help to reproduce or shed some light on it? @thirdwing @khotilov

**Comment by khotilov** [comment]
I've just tried: installing the 32 bit components of Rtools34 and R (I need the patched version since the non-patched one has a bug related to reading file time info on my system)
```
> sessionInfo()
R version 3.4.0 Patched (2017-05-08 r72665)
Platform: i386-w64-mingw32/i386 (32-bit)
Running under: Windows 10 x64 (build 14393)
```
and installing the package from within R by simply setwd into the R-package directory and running install.packages. Everything seems working.  I'll try tomorrow your long way.

**Comment by hetong007** [comment]
@khotilov I noticed it from the [win-builder service](win-builder.r-project.org). And my steps are basically how they check it.

**Comment by khotilov** [comment]
Here's what I see:
- Whichever way I install xgboost in 32bit R, it runs and passes tests with no problem.
- All checks pass for x64 when running `/c/Program\ Files/R/R-3.4.0patched/bin/x64/Rcmd.exe check --no-multiarch xgboost_0.6.4.5.tar.gz`
- When I run `/c/Program\ Files/R/R-3.4.0patched/bin/x64/Rcmd.exe check --multiarch xgboost_0.6.4.5.tar.gz` the log reports two errors for i386:
```
** running examples for arch 'i386' ... ERROR
Running examples in 'xgboost-Ex.R' failed
```
while the further printout and the log file xgboost.Rcheck/xgboost-Ex_i386.Rout show no actual errors;

as well as this:
```
** running tests for arch 'i386' ...
  Running 'testthat.R'
Warning message:
running command '"C:/PROGRA~1/R/R-34~1.0PA/bin/i386/R" CMD BATCH --vanilla  "testthat.R" "testthat.Rout"' had status 5
 ERROR
Running the tests in 'tests/testthat.R' failed.
Last 13 lines of output:
  [2]   eval-error:0.021726     train-error:0.022263
  [1]   eval-error:0.042831     train-error:0.046522
  [2]   eval-error:0.021726     train-error:0.022263
  Loading required package: Matrix
  [1]   train-error:0.046522
  [2]   train-error:0.022263
  [1]   eval-error:0.021105     train-error:0.022877
  [2]   eval-error:0.003724     train-error:0.005374
  [1]   train-error:0.261905
  testthat results ================================================================
  OK: 515 SKIPPED: 1 FAILED: 0
```
where there were no actual test errors (as well as there were no errors in the log file xgboost.Rcheck/tests_i386/testthat.Rout.fail ). The only hint is the "status 5" exit code which means "Access is denied", which I guess is due to some Win10 permissions weirdness in my system, not anything xgboost-related.

So, to summarize, I cannot reproduce your errors, and all checks seem to pass (at least I see no xgboost-related errors). Are you able to install from source and runs any other packages that require compilation in 32bit R?

**Comment by thirdwing** [comment]
I can reproduce the error in `R CMD check`, but I am not sure about the reason.

**Comment by thirdwing** [comment]
@hetong007 I successfully reproduced the error.

32-bit R will crash during the exit on the 64-bit Windows.

On my Windows Server (an Amazon EC2 instance), the `sfc /scannow` detect the `CNBJ2530.DPB` in the Windows system is corrupted.

The solution from the link below fixed my Windows server.

https://www.sysnative.com/forums/windows-update-fixes-and-tutorials/7702-win8-8-1-guide-cnbj2530-dpb-prncacla-inf-corruptions.html

**Comment by hetong007** [comment]
@thirdwing This is a great catch. Will confirm on my own Windows 10 laptop later.

The problem now becomes: how do we fix the win-builder machine of CRAN? 

Now CRAN automatically test the incoming packages through this machine first, to see if the package is generally good for further tests. If we want to update the package on CRAN, we need the Windows machine with this fix in order to pass the automated test.

**Comment by thirdwing** [comment]
OK. I see crashed due to `ntdll.dll` now. Really have no idea...

**Comment by thirdwing** [comment]
I am not very sure, but the error seems triggered by https://github.com/dmlc/xgboost/blob/master/R-package/R/xgb.train.R#L327

**Comment by thirdwing** [comment]
BTW, how about disabling test on 32-bit Windows?

**Comment by thirdwing** [comment]
Reference: https://github.com/Alexpux/MINGW-packages/issues/1104

**Comment by thirdwing** [comment]
I have uploaded the xgboost pkg for Linux/OSX/Windows to https://github.com/dmlc/drat

We can let user install from drat before we can fix this.

**Comment by khotilov** [comment]
After some time passed with some upgrades of Windows and R, now I also see the actual crashes when 32 bit R exits after running xgboost.

Disabling tests on 32-bit Windows wouldn't help, as checking the examples also crashes.

@thirdwing did you have any new ideas about it? I might try building 32 bit R with debug info to look into it.
